src/sources/bitfinex.rs

```diff
@@ -0,0 +1,132 @@
+
+use crate::non_empty::NonEmpty;
+use crate::error::Error;
+use crate::Result;
+use crate::sources::Source;
+use crate::data::{Response, Ticker, BuildRequest, PreparedRequest, Pair};
+
+use serde_json::Value;
+
+pub struct Bitfinex {
+    endpoint: String
+}
+
+
+
+impl Bitfinex {
+    pub fn new_with_endpoint(endpoint: &str) -> Bitfinex {
+        Bitfinex {
+            endpoint: endpoint.into()
+        }
+    }
+
+    pub fn new() -> Bitfinex {
+        Bitfinex::new_with_endpoint("https://api-pub.bitfinex.com")
+    }
+}
+
+fn parse_pair(s: &str) -> Result<Pair> {
+    let ch = s.chars().next().ok_or(Error::InvalidResponse("failed to parse pair".into()))?;
+    if ch != 't' {
+        return Err(Error::InvalidResponse("pair didn't start with t".into()))
+    };
+
+    let c1 = str::parse(&s[1..4])?;
+    let c2 = str::parse(&s[4..7])?;
+
+    Ok(Pair::new(c1, c2))
+}
+
+// on trading pairs (ex. tBTCUSD)
+//[
+//    SYMBOL,                  0
+//    BID,                     1
+//    BID_SIZE,                2
+//    ASK,                     3
+//    ASK_SIZE,                4
+//    DAILY_CHANGE,            5
+//    DAILY_CHANGE_RELATIVE,   6
+//    LAST_PRICE,              7
+//    VOLUME,                  8
+//    HIGH,                    9
+//    LOW                      10
+//],
+
+fn parse_bitfinex_ticker(val: &Value) -> Result<Ticker> {
+    let arr = val.as_array()
+                 .ok_or(Error::InvalidResponse("ticker value is not an array".into()))?;
+
+    if arr.len() <= 2 {
+        return Err(Error::InvalidResponse("ticker response too short".into()));
+    }
+
+    let pair_str = arr[0].as_str()
+                         .ok_or_else(|| Error::InvalidResponse("could not parse currency pair".into()))?;
+
+    let pair = parse_pair(pair_str)?;
+
+    let rate = arr[1].as_f64()
+                     .ok_or_else(|| Error::InvalidResponse("could not parse rate as f64".into()))?;
+
+    Ok(Ticker { pair, rate })
+}
+
+fn parse_bitfinex_response(arr: &Value) -> Result<Vec<Ticker>> {
+    let tvalues: Vec<Ticker> =
+        match arr {
+            Value::Array(xs) => xs.iter().map(parse_bitfinex_ticker).collect::<Result<Vec<Ticker>>>()?,
+            _ => Err(Error::InvalidResponse("invalid bitfinex response".into()))?
+        };
+    Ok(tvalues)
+}
+
+
+impl Source for Bitfinex {
+    fn name(&self) -> &str {
+        "bitfinex"
+    }
+
+    fn build_request(&self, req: BuildRequest) -> Result<PreparedRequest> {
+        let str_pairs = req
+              .pairs
+              .iter()
+              .map(|p| format!("t{}", p))
```

Comment:
I had to look at bitfinex what "t" means (thought tesnet, but it's tradig), maybe a comment